src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -223,6 +213,113 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
             return false;
         }
 
+        private static bool EncodeStringLiteralValue(string value, Span<byte> destination, out int bytesWritten)
+        {
+            if (value.Length <= destination.Length)
+            {
+                for (int i = 0; i < value.Length; i++)
+                {
+                    char c = value[i];
+                    if ((c & 0xFF80) != 0)
+                    {
+                        throw new HttpRequestException(SR.net_http_request_invalid_char_encoding);
+                    }
+
+                    destination[i] = (byte)((uint)(c));
+                }
+
+                bytesWritten = value.Length;
+                return true;
+            }
+
+            bytesWritten = 0;
+            return false;
+        }
+
+        public static bool EncodeStringLiteral(string value, Span<byte> destination, out int bytesWritten)
+        {
+            // From https://tools.ietf.org/html/rfc7541#section-5.2
+            // ------------------------------------------------------
+            //   0   1   2   3   4   5   6   7
+            // +---+---+---+---+---+---+---+---+
+            // | H |    String Length (7+)     |
+            // +---+---------------------------+
+            // |  String Data (Length octets)  |
+            // +-------------------------------+
+
+            if (destination.Length != 0)
+            {
+                destination[0] = 0; // TODO: Use Huffman encoding
+                if (IntegerEncoder.Encode(value.Length, 7, destination, out int integerLength))
+                {
+                    Debug.Assert(integerLength >= 1);
+
+                    if (EncodeStringLiteralValue(value, destination.Slice(integerLength), out int valueLength))
+                    {
+                        bytesWritten = integerLength + valueLength;
+                        return true;
+                    }
+                }
+            }
+
+            bytesWritten = 0;
+            return false;
+        }
+
+        public static bool EncodeStringLiterals(string[] values, string separator, Span<byte> destination, out int bytesWritten)
+        {
+            bytesWritten = 0;
+            if (destination.Length != 0)
+            {
+                int valueLength = 0;
+
+                if (values.Length == 0)
+                {
+                    return EncodeStringLiteral("", destination, out bytesWritten);
+                }
+                else if (values.Length == 1)
+                {
+                    return EncodeStringLiteral(values[0], destination, out bytesWritten);
+                }
```

Comment:
Move this if/else up to the very beginning of the method (before the destination.Length check)?  Otherwise we're paying for that check twice.